### PR TITLE
Issue 195: Generate mdbook and cargo documentation.

### LINKED ISSUE #195: Enable mdbook and cargo documentation.
**Problem description**
Generate mdbook and cargo documentation; Automate the same using github actions.

**Problem location**
Native client.

**Suggestions for an improvement**
Use mdbook cargo plugin for the same.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
This is a PR that uses mdbook plugin (https://rust-lang.github.io/mdBook/index.html) to generate a book on pravega-native client. It also generates the cargo documentation and published it to the github pages.

**Purpose of the change**  
Fixes #195 

**What the code does**  
Uses mdbook plugin (https://rust-lang.github.io/mdBook/index.html) to generate a book on pravega-native client. It also generates the cargo documentation and published it to the github pages.

**How to verify it**  
The book should be published on the github pages. The github action will be triggered once the PR is merged. 
Sample documentation @ https://pravega.github.io/pravega-client-rust/index.html
- Python API docs: https://pravega.github.io/pravega-client-rust/PythonBindings.html
- Cargo docs - https://pravega.github.io/pravega-client-rust/doc/pravega_client_rust/event_stream_writer/struct.EventStreamWriter.html
At the moment it can be tested using `mdbook watch ` command.

### CODE REVIEW

File: book/src/PythonBindings.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+manager=pravega_client.StreamManager("127.0.0.1:9090")
+// Create a ReaderGroup rg1 against an already created Pravega scope and Stream.
+reader_group=manager.create_reader_group("rg1", "scope", "stream")
+``` 
+
+* ## StreamWriter
+
+A StreamWriter object created using the StreamManager can be used to write events into the Pravega Stream. The events 
+that are written will appear in the Stream exactly once. The event of type String is converted into bytes with `UTF-8` encoding.
+The user can optionally specify the routing key. The StreamWriter can also write a byte arry into the Pravega Stream
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:53]:
```suggestion
The user can optionally specify the routing key. The StreamWriter can also write a byte array into the Pravega Stream 
```

File: book/src/PythonBindings.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+manager=pravega_client.StreamManager("127.0.0.1:9090")
+// lets assume the Pravega scope and stream are already created.
+writer=manager.create_writer("scope", "stream")
+// write into Pravega stream without specifying the routing key.
+writer.write_event("e1")
+// write into Pravega stream by specifying the routing key.
+writer.write_event("e2", "key1") 
+e="eventData"                                                   
+// Convert the event object to a byte array.
+e_bytes=e.encode("utf-8")
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:70]: This is not used below.

File: book/src/PythonBindings.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+A TransactionWriter created using a StreamManager  writes Events to an Event stream transactionally. All events that are 
+written as part of a transaction can be committed atomically by calling commit(). This will result in either all of those 
+events going into the stream or none of them and the commit call failing with an exception.
+
+Prior to committing a transaction, the events written to it cannot be read or otherwise seen by readers.
+
+```python  
+import pravega_client                                         
+manager=pravega_client.StreamManager("127.0.0.1:9090")        
+w1=stream_manager.create_transaction_writer(scope,"testTxn", 1)
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:87]:
```suggestion
w1 = stream_manager.create_transaction_writer(scope,"testTxn", 1)
```

File: book/src/PythonBindings.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+* ## TransactionWriter
+A TransactionWriter created using a StreamManager  writes Events to an Event stream transactionally. All events that are 
+written as part of a transaction can be committed atomically by calling commit(). This will result in either all of those 
+events going into the stream or none of them and the commit call failing with an exception.
+
+Prior to committing a transaction, the events written to it cannot be read or otherwise seen by readers.
+
+```python  
+import pravega_client                                         
+manager=pravega_client.StreamManager("127.0.0.1:9090")
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:86]:
```suggestion
manager = pravega_client.StreamManager("127.0.0.1:9090")        
```

File: book/src/PythonBindings.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+stream. The reader ensures that events returned by the stream are in order.
+
+If multiple readers are created under the same reader group then the readers co-ordinate amongst themselves and divide 
+the responsibility of reading from different segments in the Pravega stream.
+
+A sample example snippet is shown below.
+
+```python
+import pravega_client
+manager=pravega_client.StreamManager("127.0.0.1:9090")
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:119]:
```suggestion
manager = pravega_client.StreamManager("127.0.0.1:9090")
```

File: book/src/PythonBindings.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+// lets assume the Pravega scope and stream are already created.
+reader_group=manager.create_reader_group("rg1", "scope", "stream")
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:121]:
```suggestion
reader_group = manager.create_reader_group("rg1", "scope", "stream")
```

File: book/src/PythonBindings.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+reader=reader_group.create_reader("reader_id");
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:122]:
```suggestion
reader = reader_group.create_reader("reader_id");
```

File: book/src/PythonBindings.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+slice=await reader.get_segment_slice_async()
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:123]:
```suggestion
slice = await reader.get_segment_slice_async()
```

File: book/src/index.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Pravega Native Client User Guide
+
+Welcome to the Pravega RUST client user guide! This book is a companion to [Pravega RUST Client API docs](https://docs.rs/pravega_native_client).
```
Comment (tkaitchuck) [inline book/src/index.md:3]:
```suggestion
Welcome to the Pravega Rust client user guide! This book is a companion to [Pravega Rust Client API docs](https://docs.rs/pravega_native_client). 
```

File: book/src/PythonBindings.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
+# Python Bindings
+
+Pravega Native client uses [PyO3](https://pyo3.rs) to generate Python bindings.
```
Comment (tkaitchuck) [inline book/src/PythonBindings.md:3]: This should really go in a "building" section or similar as it's not really part of the API.

Comment (shrids) [reply book/src/PythonBindings.md]: ok, I have improved the doc.

File: book/src/PythonBindings.md (hunk 2 above)
Comment (shrids) [reply book/src/PythonBindings.md:69]: fixed it.

File: .github/workflows/gh.yml [hunk 11]
```diff
@@ -0,0 +1,35 @@
+name: github pages
+
+on:
+  push:
+    branches:
+      - master
+      - mdbook
```
Comment (shrids) [inline .github/workflows/gh.yml:7]:
```suggestion
```

Comment (shrids) [reply .github/workflows/gh.yml:7]: This change needs to be committed before merging the PR.

**Review by tkaitchuck — APPROVED** [review]

File: .github/workflows/gh.yml (hunk 11 above)
Comment (shrids) [reply .github/workflows/gh.yml]: done.

**Review by Tristan1900 — APPROVED** [review]

**Pull request merged**